src/Generator/Generators/CSharp/CSharpTextTemplate.cs

```diff
@@ -196,6 +198,138 @@ public void GenerateHeader()
+            public System.Runtime.InteropServices.CallingConvention callingConvention;
+
+            public override int GetHashCode()
+            {
+                return retType.GetHashCode() ^ signature.GetHashCode() ^ callingConvention.GetHashCode();
+            }
+            public override bool Equals(object obj)
+            {
+                return Equals(obj as DelegateSignature);
+            }
+            public bool Equals(DelegateSignature obj)
+            {
+                return obj != null && CompareWith(obj);
+            }
+            private bool CompareWith(DelegateSignature ds)
+            {
+                if (!retType.Equals(ds.retType) || !signature.Equals(ds.signature) || suppressWarning != ds.suppressWarning ||
+                    callingConvention != ds.callingConvention)
+                {
+                    return false;
+                }
+                return true;
+            }
+        }
+
+        private string GenerateDelegateSignature(Function function, out CSharpTypePrinterResult retType)
+        {
+            var @params = new List<string>();
+
+            TypePrinter.PushContext(CSharpTypePrinterContextKind.Native);
+
+            var retParam = new Parameter { QualifiedType = function.ReturnType };
+            retType = retParam.CSharpType(TypePrinter);
+
+            var method = function as Method;
+            var isInstanceMethod = method != null && !method.IsStatic;
+
+            if (isInstanceMethod && Options.IsMicrosoftAbi)
+            {
+                @params.Add("global::System.IntPtr instance");
+
+                if (method.IsConstructor)
+                    retType = "global::System.IntPtr";
+            }
+
+            if (!function.HasIndirectReturnTypeParameter &&
+                isInstanceMethod && Options.IsItaniumLikeAbi)
+                @params.Add("global::System.IntPtr instance");
+
+            int i = 0;
+            foreach (var param in function.Parameters)
+            {
+                if (param.Kind == ParameterKind.OperatorParameter)
+                    continue;
+
+                var typeName = param.CSharpType(TypePrinter);
+
+                @params.Add(string.Format("{0} param{1}", typeName, i++));
+
+                if (param.Kind == ParameterKind.IndirectReturnType &&
+                    isInstanceMethod && Options.IsItaniumLikeAbi)
+                    @params.Add("global::System.IntPtr instance");
+            }
+
+            if (method != null && method.IsConstructor)
+            {
+                var @class = (Class)method.Namespace;
+                if (Options.IsMicrosoftAbi && @class.Layout.HasVirtualBases)
+                    @params.Add("int " + GeneratedIdentifier("forBases"));
+            }
+
+            TypePrinter.PopContext();
+
+            return string.Join(", ", @params);
+        }
+
+        private Dictionary<DelegateSignature, string> AllDelegates = new Dictionary<DelegateSignature, string>();
```

Comment:
My idea was actually the other way around - a string as the key and your object as the value. It works this way too but with a string key the code will be much simpler and shorter because you won't need all of your code for hashes and comparison.